scripts/upload_results_to_oep.py

```diff
@@ -0,0 +1,247 @@
+
+* Create metadata (in oemetadata format, using the template in ``oemof-B3/schema/oemetadata.json``)
+  for given model results (in oemof-B3-format) and saves them locally as json files.
+* Create tables on the OpenEnergyPlatform (OEP) based on the metadata.
+* Upload model results to the OEP tables.
+* Validate metadata.
+* Upload metadata to the OEP tables.
+
+The script can delete existing tables on the OEP that have been uploaded with an affiliated OEP user
+account.
+
+The oemetadata format is a standardised json file format and is required for all data uploaded to
+the OEP. It includes the data model, the used data types, and general information about the data
+context. Tables in sqlalchemy are created based on the information in the oemetadata.
+"""
+import json
+import logging
+import os
+import pathlib
+import sys
+from datetime import date
+
+import pandas as pd
+
+from oemof_b3.config import config
+from oemof_b3.schema import oemetadata_scal, oemetadata_ts, SCHEMA_SCAL, SCHEMA_TS
+from oemof_b3.tools import data_processing as dp
+
+try:
+    from oem2orm import oep_oedialect_oem2orm as oem2orm
+except ImportError:
+    raise ImportError("Need to install oem2orm to upload results to OEP.")
+
+logger = logging.getLogger()
+
+# try to get oep_user and oep_token from .secrets.yaml
+try:
+    os.environ["OEP_USER"] = config.settings.oep_user
+    os.environ["OEP_TOKEN"] = config.settings.oep_token
+except AttributeError:
+    logger.warning(
+        "No oep_user and/or oep_token provided in oemof_b3/config/.secrets.yaml. "
+        "Will have to type provide them manually when uploading to OEP."
+    )
+
+
+SCHEMA = "model_draft"
+
+
+def save_dict_to_json(data, filepath, encoding="utf-8"):
+    with open(filepath, "w", encoding=encoding) as f:
+        return json.dump(data, f, sort_keys=False, indent=2)
+
+
+def list_diff(sample, default):
+    r"""
+    Determines extra and missing items in sample in
+    comparison with default.
+
+    Parameters
+    ----------
+    sample : list
+    default : list
+
+    Returns
+    -------
+    (extra_items, missing_items): (list, list)
+    """
+    extra_items = dp.get_list_diff(sample, default)
+
+    missing_items = dp.get_list_diff(default, sample)
+
+    if not extra_items and not missing_items:
+        return None
+
+    else:
+        return (extra_items, missing_items)
+
+
+def create_metadata(data, template=None):
+    if template is not None:
+        pass
+    elif list_diff(data.columns, SCHEMA_SCAL.columns) is None:
+        template = oemetadata_scal
+    elif list_diff(data.columns, SCHEMA_TS.columns) is None:
+        template = oemetadata_ts
+    else:
+        raise ValueError("Could not match data with the existing templates.")
+
+    metadata = template.copy()
+
+    return metadata
+
+
+def write_metadata(metadata, schema, table, title, keywords):
+    metadata["name"] = table
+    metadata["title"] = title
+    metadata["PublicationDate"] = str(date.today())
```

Comment:
The key "PublicationDate" should be "publicationDate" when using the latest oemetadata template.

Reply:
Do it later.